AOT Jit Compile Workaround

inforithmics opened the pull request:
This Pull request is a workaround for this Xamarin iOS AOT Bug.

https://github.com/xamarin/xamarin-macios/issues/12545

There are two Fixes  

1. This catches the exception and sets the Fast Mapper to null and outputs a Trace so that it can be seen in a released Version.

```
if (fastColumnSetters[i] != null) {
	try {
		fastColumnSetters[i].Invoke (obj, stmt, i);
	}
#pragma warning disable CS0618 // Type or member is obsolete
	catch (ExecutionEngineException) {
#pragma warning restore CS0618 // Type or member is obsolete
                 // Column setter has AOT Problem so don't use it.
                 fastColumnSetters[i] = null;
                 Trace.WriteLine($"FastMapper AOT Jit Exception on Type {map.MappedType.FullName} Column {cols[i].Name}");
	}
}
```

2. A Method where Fast Column Setters can be set to workround the iOS AOT Bug and still have the same-better Performance

```
SQLiteConnection.RegisterFastColumnSetter(
            typeof(CacheDto), 
            nameof(CacheDto.Date), 
            (obj, stmt, index) => { ((CacheDto)obj).Date = new DateTime(SQLite3.ColumnInt64(stmt, index)); });
```
Enhancements:
- [x] Remove the Explicit Call to SQLiteInitializers.Init() With [ModuleInitializer]
- [x] Handle Classes that are inherited from Sqlite Classes
- [x] Skip Static Attributes
- [x] Ignore Obsolete Warnings

inforithmics committed:
AOT Jit Compile Workaround

inforithmics committed:
fixed spelling

inforithmics committed:
Correct Name

inforithmics committed:
Update SQLite.cs

Read value that could not be read by default implmentation

ElteHupkes commented:
As a tangentially related remark, I'm using a similar exception catching mechanism in a different internal library, but I'm a bit worried by `ExecutionEngineException` being marked as `Obsolete`. What are the odds of a new runtime version suddenly deciding it'll throw a different kind of exception, instantly breaking everything that relies on this?

inforithmics commented:
The real workaround would be to use RegisterFastColumnSetter (I have written a SourceGenerator that Generates to code for it).

pver commented:
@praeclarum is there any chance this will be merged at some point? Using the latest sqlite-net with release builds for Xamarin.iOS is basically broken at the moment (which is a pity seeing the library is mentioned on https://learn.microsoft.com/en-us/xamarin/ios/data-cloud/data/using-sqlite-orm and the readme.md mentioning Xamarin.iOS as the initial target platform :blush: )

This PR seems like a reasonable fallback for iOS for now. And iOS library users wanting to benefit from the possible performance gains can still provide their own setter via RegisterFastColumnSetter (which could be manually written or generated like @inforithmics suggests). And even without custom fast column setters on iOS, no performance gain is also still better than not being able to use the current version at all :wink:

inforithmics committed:
Merge remote-tracking branch 'upstream/master' into FixJitCompileInIOs

inforithmics committed:
retarget Tests to .NET 6.0 Because .NET 3.1 is end of life

Dokug commented:
I would be interested to see this updated and merged, yes

inforithmics committed:
Merge remote-tracking branch 'upstream/master' into FixJitCompileInIOs

inforithmics committed:
roll forward last minor

inforithmics committed:
fixing build

inforithmics committed:
fixing global.json

inforithmics committed:
update tests

inforithmics committed:
increase version number

inforithmics committed:
fixing version

inforithmics committed:
version number

praeclarum commented:
Hello! I'm looking to get this merged for 1.10.

inforithmics committed:
Merge remote-tracking branch 'upstream/master' into FixJitCompileInIOs

inforithmics committed:
Revert Solution changes

inforithmics committed:
undo changes

inforithmics committed:
Added FastColumnSetter Unit Tests

inforithmics committed:
revert test changes

inforithmics committed:
fixed build

inforithmics commented:
I will add the Source Code Generator in a seperate pull request based on this pull request

inforithmics committed:
Add Source Generator Project

inforithmics committed:
Sqlite fist colum setters

inforithmics committed:
fixing casing

inforithmics committed:
Improrove Source Generator

inforithmics committed:
Generator improvements

inforithmics committed:
Added Unit Test

inforithmics committed:
more debugging

inforithmics committed:
Some more fixes

inforithmics committed:
Fixing Root Namespace

inforithmics committed:
Improved unit Test

inforithmics committed:
Improve column setter

inforithmics committed:
some more improvments

inforithmics committed:
More source generator fixes

inforithmics committed:
Disable debugging code

inforithmics committed:
fixing unit tests

inforithmics committed:
fixing unit test

inforithmics committed:
more unit tests

inforithmics committed:
Fixing unit tests

inforithmics committed:
More bugfixes

inforithmics committed:
disable debugging code

inforithmics committed:
fixing build

inforithmics committed:
Don't register generic types

inforithmics committed:
some fixes

inforithmics committed:
Improve properties waling

inforithmics committed:
Fixing unit test

inforithmics committed:
Adding sourcegenerator to nuget

inforithmics committed:
fixing unit test

inforithmics committed:
fixing unit test

inforithmics committed:
comment out debug output

inforithmics committed:
fixing unit tests

inforithmics committed:
don^t map readonly properties

inforithmics committed:
Don't register ignore Attributes

inforithmics committed:
Check SQLite Namespace for attributes

inforithmics committed:
A renamed test

inforithmics committed:
fix build order

inforithmics committed:
fixing make

inforithmics commented:
@praeclarum 
I have now added a SourceCodeGenerator to this pull request.
And added Unit Tests
The SourceCodeGenerator is added to the nuget package too.
I think this works for Visual Studio 2022 upwards. (CodeAnalysis.CSharp 4.0.1 and upwards).

[AssemblyDefaultNamespace].SQLiteInitializer.Init(); needs to be called so that the FastColumnSetters are initialized

Current Implementation:

All Types with TableAttribute or SqliteAttribute are added the SQLiteInitializer.Init() Function.

Possible Improvements:
Attributes that are inherited from SqliteAttributes aren't currently recognized.
```
public MyColumnAttribute : ColumnAttribute
{
}
```
MyColumnAttribute is not recognized as a SQLiteAttribute.

Classes that don't add additional Attributes but inherit from a Class that Has Sqlite Attributes are not recognized. 

```
 [Table("Test")]
 public class Test 
 {
     public int Id { get; set; }
 }

public class SecondTest : Test 
{
}

```

SecondTest is not recognized as a SQLite class.

inforithmics committed:
added decimal

inforithmics committed:
Added Timespan

inforithmics committed:
Make SqlInitializer public and initialize it only once

inforithmics committed:
fixing code generator

inforithmics committed:
fixing vulnerability warning

inforithmics committed:
rever add sqlite green

inforithmics committed:
remove debugging code

inforithmics committed:
test that no new mappers are created,.

inforithmics committed:
Add Unit Tests for all basic types

inforithmics committed:
Use ModuleInitializer for automatic Sqlite initialization

inforithmics committed:
Improved Source Generator

inforithmics committed:
cache has sqlite attribute

inforithmics committed:
fixing build

inforithmics committed:
fixing enum

inforithmics committed:
some more fixes

inforithmics committed:
fix all unit tests

inforithmics committed:
use not vulnerable mstest version

inforithmics committed:
ignore obsoelete and and don't use static or private properties

inforithmics committed:
Added Unit Tests

inforithmics committed:
more obsolete warning suppression

inforithmics committed:
added obsolete unit test

inforithmics committed:
Fixed Init and private set properties

inforithmics committed:
fixed vulnerability warning for mstest

inforithmics committed:
Reduce variables

inforithmics committed:
more variable removal

inforithmics committed:
more variable reductions

inforithmics committed:
add nullable test

inforithmics committed:
fix unit test

inforithmics committed:
improved generated code

inforithmics committed:
comment out debug output

inforithmics committed:
fixed nullable byte arrays

inforithmics committed:
added directory.build.props to solution

inforithmics committed:
Make Generated code more compact

inforithmics committed:
InitPerClass

inforithmics committed:
avoid duplicate initializations

inforithmics committed:
added partial test

inforithmics committed:
make new classes variable

inforithmics committed:
making class initializers private

inforithmics committed:
Merge remote-tracking branch 'upstream/master' into FixJitCompileInIOs

inforithmics committed:
fixing build

inforithmics committed:
some build improvements

inforithmics committed:
Fixing Build on Windows .net8-ios is deprectaed use net9-ios
